src/EnergyPlus/PCMThermalStorage.cc

```diff
@@ -0,0 +1,629 @@
+        PlantUtilities::SetComponentFlowRate(state, mUseReq, this->UseSideInletNode, this->UseSideOutletNode, this->usePlantLoc);
+        PlantUtilities::SetComponentFlowRate(state, mPlantReq, this->PlantSideInletNode, this->PlantSideOutletNode, this->sourcePlantLoc);
+
+        // Recompute heat-transfer rates using the requested flows (W)
+        // (deltaTUse/deltaTPlant were computed above; keep your existing outlet temp calcs)
+        Real64 useheatTransfer_req = mUseReq * CpWater * (useInlet.Temp - useOutletTemp);
+        Real64 plantheatTransfer_req = mPlantReq * CpWater * (plantInlet.Temp - plantOutletTemp);
+
+        // Only one side should contribute per timestep by construction; but compute net formally:
+        Real64 netPowerW = plantheatTransfer_req + useheatTransfer_req - HeatLossRate_W;
+
+        // Update stored energy (J)
+        EnergyStored += netPowerW * dt_seconds;
+
+        // Update node temps for the side that actually flowed
+        if (mUseReq > 0.0) {
+            useOutlet.Temp = useOutletTemp;
+        } else if (mPlantReq > 0.0) {
+            plantOutlet.Temp = std::min(plantOutletTemp, state.dataPlnt->PlantLoop[this->sourcePlantLoc.loopNum].MaxTemp);
+        }
+
+        // Persist rates for reporting
+        this->useheatTransfer = useheatTransfer_req;
+        this->plantheatTransfer = plantheatTransfer_req;
+
+        // Enforce bounds and update percent capacity (0–100)
+        EnergyStored = std::clamp(EnergyStored, 0.0, TankCapacity * LatentHeat);
+        PercentCapacity = 100.0 * EnergyStored / (TankCapacity * LatentHeat);
+    }
+
+    void RegisterPCMStorageOutputVariables(EnergyPlusData &state)
+    {
+        EnergyPlus::PCMStorage::PCMStorageData &PCM = EnergyPlus::PCMStorage::PCMStorageData::instance();
+
+        SetupOutputVariable(state,
+                            "Thermal Energy Storage Percent Capacity",
+                            Constant::Units::None,
+                            PCM.PercentCapacity,
+                            EnergyPlus::OutputProcessor::TimeStepType::System,
+                            EnergyPlus::OutputProcessor::StoreType::Average,
+                            PCM.Name);
+
+        SetupOutputVariable(state,
+                            "Thermal Energy Storage Heat Loss Rate",
+                            Constant::Units::W,
+                            PCM.HeatLossRate_W,
+                            EnergyPlus::OutputProcessor::TimeStepType::System,
+                            EnergyPlus::OutputProcessor::StoreType::Average,
+                            PCM.Name);
+
+        SetupOutputVariable(state,
+                            "Thermal Energy Storage Energy Stored",
+                            Constant::Units::J,
+                            PCM.EnergyStored,
+                            EnergyPlus::OutputProcessor::TimeStepType::System,
+                            EnergyPlus::OutputProcessor::StoreType::Average,
```

Comment:
This looks like a sum, as is typically done with units of joules.

        // Update stored energy (J)
        EnergyStored += netPowerW * dt_seconds;

Reply:
> This looks like a sum, as is typically done with units of joules.
> 
> ```
>     // Update stored energy (J)
>     EnergyStored += netPowerW * dt_seconds;
> ```

Usually J would be summed, but this is the total stored energy which goes up and down, so average is the correct intent.